File: org.thingml.tests/src/main/thingml/tests/Tester/launcher.py
#
# Copyright (C) 2011 SINTEF <franck.fleurey@sintef.no>
#
# Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# 	http://www.gnu.org/licenses/lgpl-3.0.txt
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from Parser import Parser
from Tester import Tester
import tempfile
import os
import re
from os import listdir
from os.path import isfile, join

#tester creates the test file from a string
#Parser gets (input,output) list from a file
def testFile(fileName):
	
	results = Parser().parse(fileName)

	
	fdump = open('../dump/'+fileName+'.dump', 'w')
	fdumpC = open('../dump/'+fileName+'C.dump', 'w')
	fdumpScala = open('../dump/'+fileName+'Scala.dump', 'w')

	#fdump.write('Fichier '+fileName+":\n")


	rootDirectory = os.getcwd()
	os.chdir(r"../../../../../../org.thingml.cmd")
	compilerDirectory = os.getcwd()

	for (a,b) in results:
		fdump.write(a+'\n'+b+'\n')
		os.chdir(rootDirectory)
		Tester().create(a)
			
		os.chdir(compilerDirectory)
		os.system("mvn clean install")
		os.system("mvn compile")
		os.chdir(rootDirectory)
		
		#!Test C
		os.chdir(compilerDirectory)
		os.system("mvn exec:java -Dexec.mainClass=\"org.thingml.cmd.Cmd\" -Dexec.args=\"c org.thingml.tests/src/main/thingml/tests/_linux/"+fileName+".thingml\"")
		bigName = fileName[0].upper()+fileName[1:]+"C"
		os.chdir("/home/thingml_out/"+bigName)
		os.system("make")
		os.system("./"+bigName)
		try:
			f = open('dump', 'r')
			res = f.readline()
			f.close()
			fdumpC.write(res+'\n')
		except IOError:
			fdumpC.write("ErrorAtCompilation\n")
		
		#!Test scala
		os.chdir(compilerDirectory)
		os.system("mvn exec:java -Dexec.mainClass=\"org.thingml.cmd.Cmd\" -Dexec.args=\"scala org.thingml.tests/src/main/thingml/tests/_scala/"+fileName+".thingml\"")

		os.chdir(tempfile.gettempdir()+"/ThingML_temp/"+fileName[0].upper()+fileName[1:])
		os.system("mvn clean package exec:java -Dexec.mainClass=\"org.thingml.generated.Main\"")
		try:
			f = open('dump', 'r')
			res = f.readline()
			f.close()
			fdumpScala.write(res+'\n')
		except IOError:
			fdumpScala.write("ErrorAtCompilation\n")

	fdump.close()
	fdumpC.close()
	fdumpScala.close()
	
	os.chdir(rootDirectory)
	
def launch():
	print(os.getcwd())
	os.chdir("..")
	mypath = "."
	onlyfiles = [ f for f in listdir(mypath) if isfile(join(mypath,f)) ]
	os.chdir("Tester")
	fileList = open("../dump/fileList.dump","w")
	for f in onlyfiles:
		match = re.match(r"(.*)\.thingml",f)
		if match is not None:
			name = re.sub(r"(.*)\.thingml",r"\1",f)
			if name != "tester":
				testFile(name)
				fileList.write(name+'\n')
	fileList.close()
				
os.chdir("../org.thingml.tests/src/main/thingml/tests/Tester/") #when called from org.thingml.tests
if not os.path.exists("../dump"):
    os.makedirs("../dump")
launch()

File: org.thingml.cep.benchmark/src/main/java/org/thingml/cep/benchmark/Launcher.java
/**
 * Copyright (C) 2014 SINTEF <franck.fleurey@sintef.no>
 *
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.thingml.cep.benchmark;


import org.sintef.thingml.*;
import org.thingml.cmd.Cmd;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @author ludovic
 */
public class Launcher {

    private  static List<Integer> nbIterList;

    private static void initList() {
        nbIterList = new ArrayList<>();
        nbIterList.add(10);
        //nbIterList.add(100000);
        //nbIterList.add(1000000000);
    }
    public static void main(String args[]) {
        if(args.length == 3) {
            initList();

            for(int i = 1; i <3; i++) {
                ThingMLModel model = Cmd.loadThingMLmodel(new File(args[i]));

                for (Integer integer : nbIterList) {
                    for (Thing thing : model.allThings()) {
                        if (thing.getName().equals("Sender")) {
                            for (Property property : thing.allProperties()) {
                                if (property.getName().equals("nbIter")) {
                                    IntegerLiteral newValue = ThingmlFactory.eINSTANCE.createIntegerLiteral();
                                    newValue.setIntValue(integer);
                                    property.setInit(newValue);
                                    break;
                                }
                            }
                            break;
                        }
                    }
                    Cmd.compile(args[0], model);

                    Configuration configuration = model.getConfigs().get(0);
                    File workingFolder = new File(System.getProperty("user.dir") + "/tmp/ThingML_Javascript/" + configuration.getName());

                    ProcessBuilder npmINstall = new ProcessBuilder("npm","install");
                    npmINstall.directory(workingFolder);

                    ProcessBuilder nodeMain = new ProcessBuilder("node","main.js");
                    nodeMain.directory(workingFolder);

                    try {
                        npmINstall.start();
                        Process process = nodeMain.start();
                        System.out.println("Execution : " + configuration.getName() + " with nbIter = " + integer);;
                        process.waitFor();

                    } catch (IOException ioex) {
                        ioex.printStackTrace();
                    } catch (InterruptedException iex) {
                        iex.printStackTrace();
                    }
                }

            }
        } else {
            System.err.println("Arguments numer not valid. \n" +
                    "Usage : [exec cmd] {javascript|arduino|linux|java} filePath1 filePath12");
        }
    }

}
